sympy/integrals/intpoly.py

```diff
@@ -0,0 +1,781 @@
+    x0 : First point on facets[index]
+    monomial_values : List of monomial values constituting the polynomial
+    monom_index : Index of monomial whose integration is being found.
+
+    >>> from sympy.abc import x, y
+    >>> from sympy.integrals.intpoly import integration_reduction_dynamic,\
+    hyperplane_parameters, gradient_terms
+    >>> from sympy.geometry.point import Point
+    >>> from sympy.geometry.polygon import Polygon
+    >>> triangle = Polygon(Point(0, 3), Point(5, 3), Point(1, 1))
+    >>> facets = triangle.sides
+    >>> a, b = hyperplane_parameters(triangle)[0]
+    >>> x0 = facets[0].points[0]
+    >>> monomial_values = [[0, 0, 0, 0], [1, 0, 0, 5],\
+                           [y, 0, 1, 15], [x, 1, 0, None]]
+    >>> integration_reduction_dynamic(facets, 0, a, b, x, (x, y), 1, 0, 1, x0,\
+                                      monomial_values, 3)
+    25/2
+    """
+    expr = S(expr)
+    value = S.Zero
+    degree = x_degree + y_degree
+    m = len(facets)
+    gens = (x, y)
+    if expr == S.Zero:
+        return expr
+
+    if not expr.is_number:
+        a, b = (S(a[0]), S(a[1])), S(b)
+
+        x_index = monom_index - max_y_degree + \
+            x_degree - 2 if x_degree >= 1 else 0
+        y_index = monom_index - 1 if y_degree >= 1 else 0
+
+        x_value, y_value =\
+            monomial_values[x_index][3], monomial_values[y_index][3]
+
+        value += x_degree * x_value * x0[0] + y_degree * y_value * x0[1]
+
+    value += left_integral(m, index, facets, x0, expr, gens)
+
+    return value/(len(dims) + degree - 1)
+
+
+def gradient_terms(binomial_power=None):
+    """Returns a list of all the possible
+    monomials between 0 and y**binomial_power
+
+    Parameters
+    ===========
+    binomial_power : Power upto which terms are generated.
+
+    Examples
+    ========
+    >>> from sympy.abc import x, y
+    >>> from sympy.integrals.intpoly import gradient_terms
+    >>> gradient_terms(2)
+    [[1, 0, 0, None], [y, 0, 1, None], [y**2, 0, 2, None], [x, 1, 0, None],\
+    [x*y, 1, 1, None], [x**2, 2, 0, None]]
+
+    """
+    terms = []
+    for x_count in range(0, binomial_power + 1):
+        for y_count in range(0, binomial_power - x_count + 1):
+            terms.append([x**x_count*y**y_count,
+                          x_count, y_count, None])
+    return terms
+
+
+def hyperplane_parameters(poly):
+    """A helper function to return the hyperplane parameters
+    of which the facets of the polygon are a part of.
+    Currently works for only 2-Polytopes.
+    Parameters
+    ==========
+    poly : The input Polygon
+
+    Examples
+    ========
+    >>> from sympy.geometry.point import Point
+    >>> from sympy.geometry.polygon import Polygon
+    >>> from sympy.integrals.intpoly import hyperplane_parameters
+    >>> hyperplane_parameters(Polygon(Point(0, 3), Point(5, 3), Point(1, 1)))
+    [((0, 1), 3), ((1, -2), -1), ((-2, -1), -3)]
+    """
+    params = []
+    vertices = list(poly.vertices)
```

Comment:
Done. I should think more about optimizing code while writing it. `append` is a more expensive operation as compared to assignment.